Create endpoint/service early to avoid unwanted create/delete volume transaction.

humblec opened the pull request:
At times, for some reason endpoint/service creation can fail in a setup. As we
    currently create endpoint/service after volume creation, later we need rollback
    of this volume transaction if endpoint/service creation failed. Considering
    endpoint/service creation is light weight, this patch promote endpoint/service
    creation to an early stage.
    
    Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/release-note-none

humblec commented:
/assign @jsafrane

humblec commented:
/retest

neolit123 commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -809,6 +809,19 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+	endpoint, service, err := p.createEndpointService(epNamespace, epServiceName, p.options.PVC.Name)
+	if err != nil {
+		klog.Errorf("failed to create endpoint/service %v/%v: %v", epNamespace, epServiceName, err)
+		return nil, 0, "", fmt.Errorf("failed to create endpoint/service %v/%v: %v", epNamespace, epServiceName, err)
```
shouldn't the error printing be left to the caller?


but instead of doing the same error twice you can also do:

```
err := fmt.Errorf("....")
klog.Error(err.Error())
return .... err
```

neolit123 reviewed (commented):
/kind bug
/priority important-longterm

please consider adding a release note if the problem can be user facing.

jsafrane commented:
/retest
/lgtm
/approve

jsafrane commented:
`FAIL: TestBootstrapControllerRoles`
`I0405 05:20:10.558] policy_test.go:258: If the change is expected, re-run with UPDATE_BOOTSTRAP_POLICY_FIXTURE_DATA=true to update the fixtures
`

humblec commented:
/retest

humblec commented:
@jsafrane Thanks !! the bazel is updated with latest patchset.. PTAL :)

humblec commented:
@jsafrane all the tests are green.. :) .. thanks .

jsafrane commented:
/lgtm
/approve

jsafrane commented:
@sig-auth-authorizers-approvers, PTAL

humblec committed:
Create endpoint/service early to avoid unwanted create/delete volume transaction.

At times, for some reason endpoint/service creation can fail in a setup. As we
currently create endpoint/service after volume creation, later we need rollback
of this volume transaction if endpoint/service creation failed. Considering
endpoint/service creation is light weight, this patch promote endpoint/service
creation to an early stage.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
Resolve merge conflict

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest

humblec commented:
@jsafrane had to rebase the PR against master. PTAL  :)

humblec commented:
/assign @liggitt

humblec commented:
@liggitt can you please review this PR ? :)

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -859,22 +872,28 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
 	}
-	epNamespace := p.options.PVC.Namespace
-	endpoint, service, err := p.createEndpointService(epNamespace, epServiceName, dynamicHostIps, p.options.PVC)
+	subset := make([]v1.EndpointSubset, 1)
```
is this re-entrant, or if the update fails, will we be left with a non-functional volume and service/endpoint?

liggitt replied:
what happens today if the service/endpoint create fails? is the function re-entrant?

humblec replied:
@liggit if the update failed, we have to cleanup the volume and endpoint. I have updated the patchset with this cleanup operation. Thanks !!

liggitt replied:
if this fails, does CreateVolume get recalled for the same PV later? if so, can it reuse the work already done, rather than trying to clean up every time? attempting to delete (like we did before with the volume, and like we're doing in this PR with the service/endpoint) does not mean we are re-entrant... network problems or process exit could leave us stranded before we have the ability to delete.

The three things this method is doing are:
* create volume
* create service
* create endpoint with volume IP

There are multiple possibilities for each of those operations:
* succeeds
* fails because it already exists
* fails for some other reason

If it already exists, are we responsible for reacting to the actual state of the already-existing thing? What does  that look like for each of the object types?

If it fails for some other reason, are we expected to fully clean up after ourselves (hint: that's not possible to guarantee... see comment about process exit or network errors), or simply return an error and get called again in the future to retry?

humblec replied:
@liggitt , previously the order was, 

- create volume
- create ep/svc 

In above workflow, we came across a limitation where eventhough the volume creation succeeded, svc creation failed due to quota setting in a customer setup . Considering volume creation and deletion is heavy compared to ep/svc creation this patch got introduced. Now, if the ep/svc creation failed or if volume creation failed , the cleanup is really light ie nothing but cleanup svc. I think the additional improvements what we can think of here is, can we make use of already created volume if there is a retry (ex:(  if ep update failed).  iic,  provision() from controller is on different go routine till it satisfies the PVC and the request is based on pvc UID. Unless user deleted and recreated the pvc object with same manifest the UID should remain same. But , the caveat here is that, heketi create a volume for each request with a different dynamically generated UID or volume name. To avoid that or to reuse what heketi already created in any of the previous iteration, we should map the volid between PVC UID and gluster/heketi vol ID ( somewhere) , such mapping does not exist. Also, as you mentioned, there are different combinations exist and we have to think through all of these. Till then, I think its better to track each transaction differently. I will also think through the possibilities of improving the workflow based on the scenarios you mentioned above, may be in a future followup design PR. :), does it sound fine ?

humblec committed:
Cleanup volume, ep/svc if endpoint update failed.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest

humblec commented:
@liggitt I have addressed your comment, PTAL :).. thanks !

humblec commented:
@liggitt Can you please take a look at this PR ? :). Thanks

liggitt commented:
queued up for review after KEP deadline today

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -809,6 +809,19 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+		epServiceName = p.provisionerConfig.customEpNamePrefix + "-" + string(p.options.PVC.UID)
+	}
+	epNamespace := p.options.PVC.Namespace
+	endpoint, service, err := p.createEndpointService(epNamespace, epServiceName, p.options.PVC)
```
rename createEndpointService to createOrGetEndpointService. since you're using the returned objects to do updates, if they already exist createEndpointService needs to fetch the current versions so we have the current resourceVersion and uid

humblec replied:
Done.. thanks!

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -859,22 +872,41 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+
+	endpoint.Subsets = subset
+	endpoint.Subsets[0].Addresses = addrlist
+	endpoint.Subsets[0].Ports = ports
```
what if other aspects of the service or endpoints are incorrect (e.g. labels are wrong, service ports are wrong)?

humblec replied:
Except `name` of ep/svc and `addrlist` field in ep object, all are  irrelevant/dummy. So, we are safe here.

humblec committed:
Rename createEndpointService() to createOrGetEndpointService()

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@liggitt addressed your comments, ptal ..  Thanks !!

humblec commented:
@liggitt I know you have many PRs to review, so please keep this too in your queue :+1: . Thanks in advance.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -859,22 +872,41 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+	endpoint.Subsets[0].Addresses = addrlist
+	endpoint.Subsets[0].Ports = ports
+
+	kubeClient := p.plugin.host.GetKubeClient()
```
move this to the top of the method to avoid doing any work if we can't get a kube client. note that returning the error here on line 888 would have left a stranded Volume (wouldn't have made it to the volume delete on line 892)

humblec replied:
Sure, Lifted to the top of the func.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -859,22 +872,41 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+
+		klog.V(3).Infof("failed to update endpoint, deleting %s", endpoint)
+
+		err = kubeClient.CoreV1().Services(epNamespace).Delete(epServiceName, nil)
```
do we also have to delete the endpoints object?

humblec replied:
The service deletion should also wipe endpoint associated with it.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -859,22 +872,41 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
+		klog.V(3).Infof("failed to update endpoint, deleting %s", endpoint)
+
+		err = kubeClient.CoreV1().Services(epNamespace).Delete(epServiceName, nil)
+		if err != nil {
```
avoid logging if the error is a NotFound error

humblec replied:
Taken care in the latest patch set.

liggitt commented:
a couple suggestions. overall, this still looks very fragile (xref https://github.com/kubernetes/kubernetes/pull/76178#discussion_r276296816), and there are plenty of scenarios which would leave a dangling service, endpoint, or volume. I don't have the context to know how severe that is, so I'll defer to sig-storage on whether the current state is acceptable.

the RBAC change lgtm

humblec committed:
Fail early if driver cant get kubeclient

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
>... I don't have the context to know how severe that is, so I'll defer to sig-storage on whether the current state is acceptable.
>
@jsafrane can you please revisit this PR ?

>
the RBAC change lgtm
>

Thanks a lot @liggitt for a detailed review and thoughts as always!!

jsafrane commented:
I thought about using owner reference for garbage collection of service + endpoints, but we would need to create PV first and then the rest "owned" by the PV. But that has shown to be problematic due to service quota. And creating a service first and then adding an owner is not robust enough either. 

So this is probably the best we can do now.

/lgtm

humblec commented:
@liggitt @jsafrane could you approve this PR ?

liggitt commented:
/approve

for `plugin/pkg/auth/authorizer` changes
